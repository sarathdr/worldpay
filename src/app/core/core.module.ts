import { APP_INITIALIZER, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ConfigService } from './services/config.service';
import { PaymentService } from './services/payment.service';
import { CartService } from './services/cart.service';
import { HttpClientModule } from '@angular/common/http';
import { FormsModule } from '@angular/forms';


export const fetchConfig = (appConfig: ConfigService) => {
  return () => appConfig.getPaymentUrls();
};

@NgModule({
  imports: [
    CommonModule,
    FormsModule
  ],
  providers: [
    ConfigService,
    PaymentService,
    HttpClientModule,
    CartService,
    {
      provide: APP_INITIALIZER,
      useFactory: fetchConfig,
      deps: [ConfigService],
      multi: true
    }
  ]
})
export class CoreModule { }
