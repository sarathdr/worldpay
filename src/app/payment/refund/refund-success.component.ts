import { Component, EventEmitter, Output } from '@angular/core';

@Component({
  selector: 'payment-refund-success',
  template: `
    <div class="content">
      <h1>Your payment has been refunded</h1>
      <p>Go to checkout to start again</p>
      <div class="row">
        <button mat-raised-button color="primary" type="button" routerLink="home">Go to home</button>
      </div>
    </div>
  `,
  styles: [
    `
          .content {
              display: flex;
              flex-direction: column;
              justify-content: space-between;
              width: 400px;
              height: 285px;
              margin: 0 auto;
          }

          .row {
              display: flex;
              flex-direction: row;
              justify-content: space-between;
          }
    `
  ]
})
export class RefundSuccessComponent {
  @Output() refundPayment = new EventEmitter<void>();
}
