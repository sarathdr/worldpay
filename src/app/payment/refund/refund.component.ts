import { Component, EventEmitter, Output } from '@angular/core';

@Component({
  selector: 'payment-refund',
  template: `
    <div class="content">
      <h1>Your have successfully made the payment</h1>
      <p>You can still refund the payment</p>
      <div class="row">
        <button mat-raised-button type="button" (click)="refundPayment.emit()">Refund</button>
        <button mat-raised-button color="primary" type="button" routerLink="/home">Go to home</button>
      </div>
    </div>
  `,
  styles: [
    `
          .content {
              display: flex;
              flex-direction: column;
              justify-content: space-between;
              width: 400px;
              height: 285px;
              margin: 0 auto;
          }

          .row {
              display: flex;
              flex-direction: row;
              justify-content: space-between;
          }
    `
  ]
})
export class RefundComponent {
  @Output() refundPayment = new EventEmitter<void>();
}
