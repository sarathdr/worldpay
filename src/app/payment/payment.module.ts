import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AuthorizeComponent } from './authorize/authorize.component';
import { AuthorizeContainer } from './authorize/authorize.container';
import { ConfirmComponent } from './confirm/confirm.component';
import { ConfirmContainer } from './confirm/confirm.container';
import { RefundComponent } from './refund/refund.component';
import { RefundContainer } from './refund/refund.container';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { MatButtonModule, MatFormFieldModule, MatInputModule, MatSelectModule } from '@angular/material';
import { AuthorizeFormComponent } from './authorize/authorize-form.component';
import { ConfirmPaymentGuard } from './guards/confirm-payment.guard';
import { ExpiryDateModule } from '../core/directives/expiry-date.module';
import { RefundPaymentGuard } from './guards/refund-payment.guard';
import { RefundSuccessComponent } from './refund/refund-success.component';

@NgModule({
  declarations: [
    AuthorizeComponent,
    AuthorizeFormComponent,
    AuthorizeContainer,
    ConfirmComponent,
    ConfirmContainer,
    RefundComponent,
    RefundSuccessComponent,
    RefundContainer
  ],
  imports: [
    CommonModule,
    FormsModule,
    RouterModule.forChild([
      {
        path: '',
        component: AuthorizeContainer
      },
      {
        path: 'confirm',
        component: ConfirmContainer,
        canActivate:[ConfirmPaymentGuard],
        canDeactivate:[ConfirmPaymentGuard]
      },
      {
        path: 'refund',
        component: RefundContainer,
        canActivate:[RefundPaymentGuard],
        canDeactivate:[RefundPaymentGuard]
      }
    ]),
    MatFormFieldModule,
    MatInputModule,
    MatSelectModule,
    ReactiveFormsModule,
    MatButtonModule,
    ExpiryDateModule
  ],
  providers:[
    ConfirmPaymentGuard,
    RefundPaymentGuard
  ]
})
export class PaymentModule { }
