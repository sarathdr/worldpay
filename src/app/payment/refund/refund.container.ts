import { Component } from '@angular/core';
import { PaymentService } from '../../core/services/payment.service';
import { Router } from '@angular/router';
import { Subscription } from 'rxjs';
import { take } from 'rxjs/operators';

@Component({
  selector: 'payment-refund-page',
  template: `
    <payment-refund (refundPayment)="refund()" *ngIf="!refundSuccess"></payment-refund>
    <payment-refund-success *ngIf="refundSuccess"></payment-refund-success>
  `
})
export class RefundContainer {
  refundSuccess = false;
  constructor(private paymentService: PaymentService,
              private router: Router) {
  }

  refund() {
    this.paymentService.refundPayment()
      .pipe(take(1))
      .subscribe(() => this.refundSuccess = true);
  }
}
