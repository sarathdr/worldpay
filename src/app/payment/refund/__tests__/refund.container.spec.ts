import { TestBed } from '@angular/core/testing';
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { MatButtonModule, MatListModule, MatTableModule } from '@angular/material';
import { RouterTestingModule } from '@angular/router/testing';
import { PaymentService } from '../../../core/services/payment.service';
import { CartService } from '../../../core/services/cart.service';
import { By } from '@angular/platform-browser';
import { of } from 'rxjs';
import { Router } from '@angular/router';
import { RefundContainer } from '../refund.container';


describe('RefundContainer', () => {

  let fixture;
  let refundPayment: jest.Mock;
  let navigate: jest.Mock;

  beforeEach(() => {
    refundPayment = jest.fn();
    navigate = jest.fn();

    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        MatListModule,
        MatButtonModule,
        MatTableModule
      ],
      declarations: [
        RefundContainer,
        MockPaymentRefund,
        MockPaymentRefundSuccess
      ],
      providers: [
        CartService,
        {
          provide: PaymentService,
          useValue: {
            refundPayment
          }
        },
        {
          provide: Router,
          useValue: {
            navigate
          }
        }
      ]
    });

    fixture = TestBed.createComponent(RefundContainer);
  });

  it('should render the component correctly', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should refund payment', () => {
    fixture.detectChanges();

    refundPayment.mockReturnValue(of({}))
    const refundPaymentComponent = fixture.debugElement.query(By.directive(MockPaymentRefund)).componentInstance;
    refundPaymentComponent.refundPayment.emit();
    fixture.detectChanges();

    expect(refundPayment).toHaveBeenCalled();
    expect(fixture).toMatchSnapshot();
  });

  @Component({
    selector: 'payment-refund',
    template: `
      <div>Refund: {{ cart | json }}</div>
    `
  })
  class MockPaymentRefund {
    @Input() cart;
    @Output() refundPayment = new EventEmitter<void>();
  }

  @Component({
    selector: 'payment-refund-success',
    template: `
      <div>Refund success</div>
    `
  })
  class MockPaymentRefundSuccess {
  }

});
