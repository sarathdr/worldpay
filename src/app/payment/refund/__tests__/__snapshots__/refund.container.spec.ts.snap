// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefundContainer should refund payment 1`] = `
<payment-refund-page
  paymentService={[Function Object]}
  refundSuccess={[Function Boolean]}
  router={[Function Object]}
>
  <payment-refund-success>
    <div>
      Refund success
    </div>
  </payment-refund-success>
</payment-refund-page>
`;

exports[`RefundContainer should render the component correctly 1`] = `
<payment-refund-page
  paymentService={[Function Object]}
  refundSuccess="false"
  router={[Function Object]}
>
  <payment-refund>
    <div>
      Refund: 
    </div>
  </payment-refund>
</payment-refund-page>
`;
